File: 프로그래머스/unrated/160586. 대충 만든 자판/대충 만든 자판.py
from collections import defaultdict

def solution(keymap, targets):
    # 0으로 초기화
    sH = defaultdict(int)
    answer = []

    # ex) "ABACD"
    for key in keymap:
        # ex) "A"
        for i, char in enumerate(key):
            # 기존 눌러야할 횟수보다 적을 경우에만 변경
            if sH[char] > (i+1) or sH[char] == 0:
                # ex) "A" : 1
                sH[char] = i + 1

    # ex) "ABCD"
    for key in targets:
        point = 0

        # ex) "A"
        for char in key:
            if char in sH:
                point += sH[char]
            else:
                point = -1
                break

        answer.append(point)

    return answer

File: 프로그래머스/1/160586. 대충 만든 자판/대충 만든 자판.java
import java.util.*;

class Solution {
    public int[] solution(String[] keymap, String[] targets) {
        int[] answer = new int[targets.length];
        HashMap<String, Integer> map = new HashMap<>();
        
        for(String key : keymap){
            String[] keySplit = key.split("");
            for(int i = 0; i < keySplit.length; i++){
                if(map.get(keySplit[i]) != null){
                    if(map.get(keySplit[i]) > i+1){
                        map.replace(keySplit[i], i+1);
                    }
                } else{
                    map.put(keySplit[i], i+1);
                }
            }
        }
        
        for(int i = 0; i < targets.length; i++){
            int sum = 0;
            for(String str : targets[i].split("")){
                if(map.containsKey(str)){
                    sum += map.get(str);
                } else{
                    sum = -1;
                    break;
                }
            }
            answer[i] = sum;
        }
        
        return answer;
    }
}